File: nicktelsan/day03/powerUsage.js
import fs from 'fs'
import path from 'path'

const __dirname = path.resolve()

function mostCommonBit(array, position) {
  let zero = 0
  let one = 0

  array.forEach(element => {
    element[position] === '0' ? zero++ : one++
  })

  return zero > one ? '0' : '1'
}

function flipBits(number) {
  let newNumber = ''

  for (var i = 0; i < number.length; i++) {
    newNumber += number[i] === '0' ? '1' : '0'
  }
  return newNumber
}

function main() {
  const input = fs.readFileSync(path.join(__dirname, 'day03/input'), 'utf8')
  const inputArray = input.split('\n')
  const bitLength = inputArray[0].length

  let gamma = ''

  for (var i = 0; i < bitLength; i++) {
    gamma += mostCommonBit(inputArray, i)
  }

  let epsilon = flipBits(gamma)

  console.log(gamma)
  gamma = parseInt(gamma, 2)
  epsilon = parseInt(epsilon, 2)

  console.log(`gamma: ${gamma}`)
  console.log(`epsilon: ${epsilon}`)
  console.log(`Power: ${gamma * epsilon}`)
}

main()

File: nicktelsan/day03/powerUsage.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

func mostCommonBit(array []string, position int) string {
	zero := 0
	one := 0

	for i := 0; i < len(array); i++ {
		if array[i][position] == '0' {
			zero++
		} else {
			one++
		}
	}

	if zero > one {
		return "0"
	} else {
		return "1"
	}
}

func flipBits(number string) string {
	newNumber := ""

	for i := 0; i < len(number); i++ {
		if string(number[i]) == "0" {
			newNumber += "1"
		} else {
			newNumber += "0"
		}
	}
	return newNumber
}

func main() {
	content, err := ioutil.ReadFile("day03/input")

	if err != nil {
		log.Fatal(err)
	}

	inputArray := strings.Split(string(content), "\n")
	bitLength := len(inputArray[0])

	gamma := ""

	for i := 0; i < bitLength; i++ {
		gamma += mostCommonBit(inputArray, i)
	}

	epsilon := flipBits(gamma)

	decimalGamma, _ := strconv.ParseInt(gamma, 2, 64)
	decimalEpsilon, _ := strconv.ParseInt(epsilon, 2, 64)
	power := decimalGamma * decimalEpsilon

	fmt.Printf("gamma: %v\n", decimalGamma)
	fmt.Printf("epsilon: %v\n", decimalEpsilon)
	fmt.Printf("Power: %v\n", power)

}
